include/tvm/internal/RangeCounting.h

```diff
@@ -0,0 +1,121 @@
+/* Copyright 2017-2020 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#pragma once
+
+#include <tvm/api.h>
+
+#include <tvm/Range.h>
+
+#include <list>
+
+namespace tvm::internal
+{
+/** Keep track of the union and (set) subtraction of integer ranges, accounting
+ * for the number of times a number was added.
+ * For example adding the set {1,2,3,4} (represented by Range(1,4)) and the set
+ * {3,4,5,6} (Range(3,4)) will result in a single range, Range(1,6). But
+ * internally, it is known that 3 and 4 appeared twice so that removing {2,3,4}
+ * (Range(2,3)) will result in a set {1,3,4,5,6}, i.e. the ranges Range(1,1) and
+ * Range(3,4).
+ *
+ *
+ * Internally, the ranges are represented by their limits, following the idea in
+ * https://stackoverflow.com/a/32869470/11611648, e.g. Range(1,4) = {1,2,3,4} is
+ * represented by {(1,+), (5,-)} where the lower limit (opening, denoted by +)
+ * is included and the upper limit (closing, denoted by -) is excluded.
+ * A list of such limits is kept as a representation of the current state.
+ * The number of appearances of an integer is encoded by the number of + and -
+ * appearing to reach that number from the start of this list. For example, if
+ * the list is {(1,+), (3,+), (5,-), (6,+), (7,-), (7,-), (9,+), (10,-)}, all
+ * number up to 0 don't appear, 1 and appear once, 3 and 4 appear twice, 5
+ * appears once, 6 appears twice, 7 and 8 don't appear, 9 appears once and all
+ * numbers from 10 don't appear. The resulting range representation is
+ * { Range(1,7), Range(9,1) }.
+ */
```

Comment:
I think the documentation should add a note about what's the motivation/usage for this class in the framework

Reply:
Done in 1bea6bb15bcd45398656e2ad06cc6edf68575c82